### ISSUE TITLE
[Performance impacting] Upgrade Android Build rules to use standard android library Gradle plugin

### ISSUE THREAD

**Issue by metalhead8816** [Kotlin/kotlinx.coroutines#4158]
The current android build rules [relies on a Google Android maven artifact](https://github.com/Kotlin/kotlinx.coroutines/blob/master/ui/kotlinx-coroutines-android/build.gradle.kts#L17) which has not been updated in ~8 years. This is very deprecated as described [here](https://stackoverflow.com/questions/34338628/android-maven-plugin-sdk-version)

Due to this very old Android version, coroutines cannot access newer Android APIs directly like Handler.createAsync. These methods are still used in coroutines today, but [require expensive reflection to access](https://github.com/Kotlin/kotlinx.coroutines/blob/master/ui/kotlinx-coroutines-android/src/HandlerDispatcher.kt#L86-L89). This impacts startup performance for all coroutines users on Android.

Ideally moving to standard android library Gradle plugin can be done to allow access to newer Android APIs.

**Comment by dkhalanskyjb** [comment]
I've looked into the issue, and solving this doesn't seem straightforward.

* Android artifacts newer than the version we use seemingly require the Android Gradle plugin.
* The Android Gradle plugin interacts with the Kotlin plugin, and the Kotlin plugin requires the Android plugin to be at least 7.3.1. That's a jump of three major versions! The effect is hard to predict, and we'd need to manually check the published artifacts after the migration.
* The Kotlin plugin for Android is separate from the Kotlin plugin for the JVM. I don't know why that is.
* Our existing build scripts that expect to see either the `kotlin-jvm` or `kotlin-multiplatform` plugin start failing when `kotlin-android` is introduced, and fixing it is not purely syntactic manipulation and requires research. For example, `kotlin.target.compilations` is empty for `KotlinAndroidExtension` for some reason. After changing this, we'd also probably need to verify that unrelated modules didn't break in the process.

Basically, it looks like this migration can't be performed without diving deep, learning about the interactions between Android and Kotlin plugins, performing a set of non-trivial changes, and carefully analyzing the potential impact.

**Comment by JakeWharton** [comment]
You can switch to using https://search.maven.org/artifact/org.robolectric/android-all as your `compileOnly` dependency for the Android APIs as an easy fix for gaining access to newer APIs.
